onnxruntime/test/shared_lib/test_inference.cc

```diff
@@ -305,6 +308,99 @@ TEST(CApiTest, custom_op_handler) {
 #endif
 }
 
+template <typename T>
+void custom_mul(const T* X, const T* Y, T* Z, int64_t size) {
+#ifdef USE_CUDA
+  cuda_mul(X, Y, Z, size);
+#else
+  for (int64_t i = 0; i < size; i++) {
+    Z[i] = X[i] * Y[i];
+  }
+#endif
+}
+
+struct VariedInputCustomOpKernel {
+  VariedInputCustomOpKernel(Ort::CustomOpApi ort, const OrtKernelInfo* /*info*/) : ort_(ort) {
+  }
+
+  void Compute(OrtKernelContext* context) {
+    // Setup inputs
+    const OrtValue* input_X = ort_.KernelContext_GetInput(context, 0);
+    const OrtValue* input_Y = ort_.KernelContext_GetInput(context, 1);
+    OrtTensorTypeAndShapeInfo* input_X_info = ort_.GetTensorTypeAndShape(input_X);
+    ONNXTensorElementDataType input_X_type = ort_.GetTensorElementType(input_X_info);
+    OrtTensorDimensions dimensions(ort_, input_X);
+    // Setup output
+    OrtValue* output = ort_.KernelContext_GetOutput(context, 0, dimensions.data(), dimensions.size());
+    OrtTensorTypeAndShapeInfo* output_info = ort_.GetTensorTypeAndShape(output);
+    int64_t size = ort_.GetTensorShapeElementCount(output_info);
+    ort_.ReleaseTensorTypeAndShapeInfo(output_info);
+    // Compute
+    switch (input_X_type) {
+      case ONNX_TENSOR_ELEMENT_DATA_TYPE_FLOAT:
+        custom_mul(ort_.GetTensorData<float>(input_X),
+                   ort_.GetTensorData<float>(input_Y),
+                   ort_.GetTensorMutableData<float>(output), size);
+        break;
+      case ONNX_TENSOR_ELEMENT_DATA_TYPE_DOUBLE:
+        custom_mul(ort_.GetTensorData<double>(input_X),
+                   ort_.GetTensorData<double>(input_Y), 
+                   ort_.GetTensorMutableData<double>(output), size);
+        break;
+      default:
+        ORT_THROW("Unsupported input type: ", input_X_type);
+    }
+  }  // Compute
+
+ private:
+  Ort::CustomOpApi ort_;
+};
+
+struct VariedInputCustomOp : Ort::CustomOpBase<VariedInputCustomOp, VariedInputCustomOpKernel> {
+  explicit VariedInputCustomOp(const char* provider) : provider_(provider) {}
+  void* CreateKernel(Ort::CustomOpApi api, const OrtKernelInfo* info) const { return new MyCustomKernel(api, info); };
+  const char* GetName() const { return "CustomMul"; };
+
+  const char* GetExecutionProviderType() const { return provider_; };
+
+  size_t GetInputTypeCount() const { return 2; };
+  ONNXTensorElementDataType GetInputType(size_t /*index*/) const { return ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED; };
```

Comment:
How does one know what types are supported by a custom kernel? We plan to have a public repository of custom kernels such that users can compile them and use them with ORT. I assume these users will like to see the schema of a custom kernel before compiling it in. "Undefined" can be super confusing.

Reply:
The custom kernels we will publish has all fixed type right? Namely, for each input and output type, the type are all precisely assigned to a type like integer, float or double, there will be no confusion at all.
The fix here is trying to cancel the behavior that we give each input and output a fixed type, it enhance it by saying that "ok, if you are sure that this type could be of any type, let it be undefined", make sense?

Reply:
How about ONNX_TENSOR_ELEMENT_DATA_TYPE_ANY with a comment saying this is used in the context of custom ops only?

Reply:
When a user implements a custom op, s/he definitely is sure about what's types would be supported, since there is no magic to support a new type without adding some code.

the issue on std::vector<ONNXTensorElementDataType> GetInputType(...) const {...} solution could be fixed by a const C array ending with ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED.

Reply:
I had some discussion with Ryan and Randy. We don't want to complicate the API. We can use ONNX_TENSOR_ELEMENT_DATA_TYPE_UNDEFINED and document it in the custom ops API struct clearly to say what it means.